subsys/net/openthread/rpc/server/ot_rpc_dns_client.c

```diff
@@ -0,0 +1,763 @@
+}
+
+static void address_response_callback(otError error, const otDnsAddressResponse *response,
+				      void *context, uint32_t callback_slot)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, ((2 * sizeof(uint32_t)) + 2) +
+					   ((2 * sizeof(uintptr_t)) + 2));
+
+	nrf_rpc_encode_uint(&ctx, (uint32_t) error);
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)response);
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)context);
+	nrf_rpc_encode_uint(&ctx, callback_slot);
+
+	nrf_rpc_cbor_cmd_no_err(&ot_group, OT_RPC_CMD_DNS_ADDRESS_RESPONSE_CB, &ctx,
+				ot_rpc_decode_void, NULL);
+}
+
+
+NRF_RPC_CBKPROXY_HANDLER(address_response_callback_encoder, address_response_callback,
+			 (otError error, const otDnsAddressResponse *response, void *context),
+			 (error, response, context));
+
+static void browse_response_callback(otError error, const otDnsBrowseResponse *response,
+				     void *context, uint32_t callback_slot)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, ((2 * sizeof(uint32_t)) + 2) +
+					   ((2 * sizeof(uintptr_t)) + 2));
+
+	nrf_rpc_encode_uint(&ctx, (uint32_t) error);
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)response);
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)context);
+	nrf_rpc_encode_uint(&ctx, callback_slot);
+
+	nrf_rpc_cbor_cmd_no_err(&ot_group, OT_RPC_CMD_DNS_BROWSE_RESPONSE_CB, &ctx,
+				ot_rpc_decode_void, NULL);
+}
+
+NRF_RPC_CBKPROXY_HANDLER(browse_response_callback_encoder, browse_response_callback,
+			 (otError error, const otDnsBrowseResponse *response, void *context),
+			 (error, response, context));
+
+static void service_response_callback(otError error, const otDnsServiceResponse *response,
+				     void *context, uint32_t callback_slot)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, ((2 * sizeof(uint32_t)) + 2) +
+					   ((2 * sizeof(uintptr_t)) + 2));
+
+	nrf_rpc_encode_uint(&ctx, (uint32_t) error);
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)response);
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)context);
+	nrf_rpc_encode_uint(&ctx, callback_slot);
+
+	nrf_rpc_cbor_cmd_no_err(&ot_group, OT_RPC_CMD_DNS_SERVICE_RESPONSE_CB, &ctx,
+				ot_rpc_decode_void, NULL);
+}
+
+NRF_RPC_CBKPROXY_HANDLER(service_response_callback_encoder, service_response_callback,
+			 (otError error, const otDnsServiceResponse *response, void *context),
+			 (error, response, context));
+
+static void resolve_address(enum ot_rpc_cmd_server cmd, const struct nrf_rpc_group *group,
+			    struct nrf_rpc_cbor_ctx *ctx)
+{
+	struct zcbor_string hostname;
+	otDnsQueryConfig config;
+	otDnsAddressCallback cb;
+	otError error = OT_ERROR_FAILED;
+	void *cb_ctx;
+	bool decoded_config;
+
+	zcbor_tstr_decode(ctx->zs, &hostname);
+	cb = (otDnsAddressCallback)nrf_rpc_decode_callbackd(ctx, address_response_callback_encoder);
+	cb_ctx = (void *)nrf_rpc_decode_uint(ctx);
+
+	ot_rpc_decode_dns_query_config(ctx, &decoded_config, &config);
+
+	if (!nrf_rpc_decoding_done_and_check(group, ctx)) {
```

Comment:
Once you call `nrf_rpc_decoding_done`, `hostname` will point to a released memory. Therefore, you have to copy this to another buffer before proceeding or call the OT API before calling `nrf_rpc_decoding_done`.

Reply:
Good catch, fixed.

Reply:
Fixed also other occurrences in this file